Do not release queries in Get/Select. Fix #25.

jgiles opened the pull request:
Releasing query objects in Get/Select could easily lead to
double-releases, which can cause dangerous and tricky data races.

jgiles commented:
It would probably also make sense to remove ReleaseQuery entirely (it's hard to explain how it should fit into the query lifecycle). But, that's a breaking API change.

mmatczuk commented:
> It would probably also make sense to remove ReleaseQuery entirely (it's hard to explain how it should fit into the query lifecycle). But, that's a breaking API change.

`ReleaseQuery` could easily be made private no need to export it.

> Releasing query objects in Get/Select could easily lead to double-releases, which can cause dangerous and tricky data races.

That's true, but I'd rather keep the behavior not to break existing code.

We could:

* introduce `gocqlx.AutoRelease` flag set to `true` by default
* introduce `Iterx.release` flag and a setter for it `Release(release bool)`
* release only when the flag is true
* rename `ReleaseQuery` to `releaseQuery`

WDYT?

jgiles commented:
> `ReleaseQuery` could easily be made private no need to export it.

It's possible there are clients who are currently calling the method, but if we don't care about breaking them we can remove it.

> That's true, but I'd rather keep the behavior not to break existing code.

Removing the release isn't going break anyone's code. At worst they will face a minor performance regression in the form of slightly increased GC pressure.

On the other hand, continuing to do the release is "dangerous by default" - it has caused and will continue to cause application correctness bugs. Bugs of the worst kind - sporadic, non-reproducible, and difficult to track down even with the Go race detector (because the races don't directly involve the read/write operations in the ReleaseQuery call, instead cropping up between reads/writes of two later goroutines when the double-released objects are re-issued from the `sync.Pool`).

And on the performance regression... recently I've begun to doubt whether the gocql query pooling is even worth it. The number unpooled of allocations involved in actual query binding/execution/scanning may dwarf the benefits of pooling the query objects. At some point I want to put together a benchmark... gocql doesn't seem to have one, which itself is a red flag for a performance optimization feature.

jgiles commented:
Agreed on removing the query field of iterx entirely; you can't reliably check against use of already-released queries anyway (client could release it directly).

I've added a commit that does this. It's a breaking API change because the ReleaseQuery method is removed. If that makes sense, I can squash these.

mmatczuk commented:
Please squash it, also if you could update examples and readme with defer release on query it would be great!

jgiles commented:
Squashed and updated the docs.

Even the `ExecRelease` utility is beginning to look a little dangerous to me, but at least it is very explicit about what it does.

mmatczuk commented:
Thanks!
